### ISSUE TITLE
Find the right Schema integration mechanism

### ISSUE THREAD

**Issue by rsdoiel** [rsdoiel/sqlish#19]
Will a schema integration make a plugin system better?  Can it be optional and still useful or would it need to be required? How would it work across joins?

Yawn's suggestion from release 0.0.5:

> mh i was thinking.. what if you could load a db schema and therefore validation would be available 
> also on insert and stuff? 
> 
> something like

``` javascript
sqlish.define('table', { 
  row: Integer({ length: 32})
});
sqlish.createTable('table');
sqlish.insert('table',{
  row: 3,
  row2: 4 // this gets ignored
});

```

> it's just an example.. i think this should be considered carefully to improve security but without 
> sacrificing usability

**Comment by yawnt** [comment]
i was thinking, instead of schemas which are fixed and not flexible, we could implement hooks and let the user do the filtering.. something like

``` javascript
sql.on('table', function(value) {
  // filtering

  return sane_value;
}); // gets called where 'table' is used
```

this would help in a future abstraction for mongo etc

**Comment by rsdoiel** [comment]
I like your approach with on().  Would the function parameter handle all
columns or would you do something like

```
sql.on('some_table', 'some_column', function (err, val) {
    ...
});
```

I your version with fewer parameters but I'm wondering if it would allow
for more incremental composition if we also support the second.  We
could check the argument count and assume 2 args works across the whole
table, 3 args works only on a column (or doc attribute in NoSQL).  That
way if you wanted to use sqlish to do your application specific validation
you could without requiring a whole full blown validation system.

On a tangential note, I've been thinking about eventual dialect support.
Rather than embedding dialect in each function as I've done we would have
each dialect define functions that worked differently and fallback to
something common for everything else (e.g. fallback to SQL 92 to be safe).
Building up
current PosgreSQL support (or MySQL support) would then be an issue of
defining differences and leaving the fallbacks for everything else.
Does that make sense?

```
// plugin type, plugin name, plugin functions which overwrite default
```

behavior
    sql.plugin("Dialect", "SQL_CRAZINESS_2012", {
        select: function () {
           ...
        },
        form: function () {
           ...
        }, ...
    });

That would allow splitting a dialects beyond our initial 3 into a
sub-modules or 3rd party modules.  What do you think?

Thanks, Robert

**Comment by yawnt** [comment]
about plugins:

yeah that's how i made it in my crappy (because i was focused on the module itself) sql builder of my orm (https://github.com/yawnt/hater) .. also the plugin should monkey patch the current implementation of the core to provide consistency

eg: select would try first plugin.select and then fallback to core.select

about the hooks:

i like the table approach because otherwise you'd have to check (and store) more functions.. i'd rather keep an api consistent than providing multiple options that might confuse the user (think of the major rewrite the mongoose team had to do of their apis for v3)

update:

we could provide hooks for specific methods and a 'run-every-time' hook for the table to improve flexibility

``` javascript
.on('insert<tableName>'... ) => // would run '<tableName>' and 'insert<tableName>' 
```

ps: please use the code tag in markdown because it makes reading a lot easier :P

**Comment by rsdoiel** [comment]
That makes sense. Let's stick with your approach. Once I finish adding
transaction, make code passes jslint and nail obvious bugs as I review the
code I think we're ready to release 0.0.5. We can re-factor for plugins and
add your hooks together in 0.0.6.  I also want to think a little about
keeping the learning curse shallow. Partly because it will make writing the
docs easier but I also think it is important if this is going to be of more
use to others.

Thanks,

Robert

**Comment by yawnt** [comment]
i can do the jslint, cleanup code part .... i think it's better if you do the architecture part because you are the owner of the lib

ps: why 4 spaces?

**Comment by rsdoiel** [comment]
I've bumped the tag to 0.0.5. I've cleaned up the internal model a bit. It is still messy but I need to use it in the real world to have a better since of what will really work. I added a define method. It's crude but it lets you replace verbs, clauses, toString() and a new things called execute().  I'm guessing that execute() will be handy if you want to either integrate sqlish with a db connection library or you're running it in the Mongo shell and want to generate executable Mongo commands from the internal sqlish structure.

4 spaces because I haven't bothered to tweak the various editors I'm using.  I actually prefer tabs myself but I'm constantly bouncing between editors (SubEthaEdit and Brackets on macs and Cloud 9 on my Chromebook), and several different vi/emacs combos depending on the vm/host I'm working on remotely.  I really should sit down and get all the different configurations in sync but I'm lazy :-)
